pkg/cosign/verify.go

```diff
@@ -84,141 +84,153 @@ func (w *reverseDSSEVerifier) VerifySignature(s io.Reader, m io.Reader, opts ...
+	// Now verify the cert, then the signature.
+	if err := TrustedCert(cert, co.RootCerts); err != nil {
+		return nil, err
+	}
+	if co.CertEmail != "" {
+		emailVerified := false
+		for _, em := range cert.EmailAddresses {
+			if co.CertEmail == em {
+				emailVerified = true
+				break
+			}
+		}
+		if !emailVerified {
+			return nil, errors.New("expected email not found in certificate")
 		}
 	}
+	return verifier, nil
+}
 
-	se, err := ociremote.SignedEntity(signedImgRef, co.RegistryClientOpts...)
+func tlogValidatePublicKey(rekorClient *client.Rekor, pub crypto.PublicKey, b64sig string, payload []byte) error {
+	pemBytes, err := cryptoutils.MarshalPublicKeyToPEM(pub)
 	if err != nil {
-		return nil, false, err
+		return err
 	}
-	// Both of the SignedEntity types implement Digest()
-	h, err := se.(interface{ Digest() (v1.Hash, error) }).Digest()
+	_, _, err = FindTlogEntry(rekorClient, b64sig, payload, pemBytes)
+	return err
+}
+
+func tlogValidateCertificate(rekorClient *client.Rekor, cert *x509.Certificate, b64sig string, payload []byte) error {
+	pemBytes, err := cryptoutils.MarshalCertificateToPEM(cert)
 	if err != nil {
-		return nil, false, err
+		return err
+	}
+	uuid, _, err := FindTlogEntry(rekorClient, b64sig, payload, pemBytes)
+	if err != nil {
+		return err
+	}
+	// if we have a cert, we should check expiry
+	// The IntegratedTime verified in VerifyTlog
+	e, err := GetTlogEntry(rekorClient, uuid)
+	if err != nil {
+		return err
+	}
+	return checkExpiry(cert, time.Unix(*e.IntegratedTime, 0))
+}
+
+// VerifySignatures does all the main cosign checks in a loop, returning the verified signatures.
+// If there were no valid signatures, we return an error.
+func VerifyImageSignatures(ctx context.Context, signedImgRef name.Reference, co *CheckOpts) (checkedSignatures []oci.Signature, bundleVerified bool, err error) {
+	// Enforce this up front.
+	if co.RootCerts == nil && co.SigVerifier == nil {
+		return nil, false, errors.New("one of verifier or root certs is required")
 	}
 
 	// TODO(mattmoor): We could implement recursive verification if we just wrapped
 	// most of the logic below here in a call to mutate.Map
 
-	sigs, err := accessor(se)
+	se, h, err := getSignedEntity(signedImgRef, co.RegistryClientOpts)
+	if err != nil {
+		return nil, false, err
+	}
+	sigs, err := se.Signatures()
 	if err != nil {
 		return nil, false, err
 	}
 	sl, err := sigs.Get()
 	if err != nil {
 		return nil, false, err
 	}
+
+	validationErrs := []string{}
+
+	var rekorClient *client.Rekor
+	if co.RekorURL != "" {
+		rekorClient, err = rekor.GetRekorClient(co.RekorURL)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+
 	for _, sig := range sl {
 		if err := func(sig oci.Signature) error {
-			b64sig, err := sig.Base64Signature()
-			if err != nil {
-				return err
-			}
-			payload, err := sig.Payload()
-			if err != nil {
-				return err
-			}
-			cert, err := sig.Cert()
+			b64sig, payload, cert, err := unpackOCISignature(sig)
```

Comment:
We should kill this function, have the stuff below take the `oci.Signature`, and access the pieces they care about.